tracing-subscriber/src/registry/fmt.rs

```diff
@@ -0,0 +1,349 @@
+};
+use smallvec::SmallVec;
+use std::{cell::RefCell, fmt, io, marker::PhantomData};
+use tracing_core::{
+    span::{Attributes, Id, Record},
+    Event, Subscriber,
+};
+
+/// A [Layer] that logs formatted representations of `tracing` events.
+pub struct FmtLayer<
+    S = Registry,
+    N = format::DefaultFields,
+    E = format::Format<format::Full>,
+    W = fn() -> io::Stdout,
+> {
+    // TODO(david): don't force boxing here. consider:
+    // - starting on https://github.com/tokio-rs/tracing/issues/302
+    // - making it a generic param; defaulting to a boxed impl.
+    // - rename this, because this isn't per-subscriber filtering.
+    is_interested: Box<dyn Fn(&Event<'_>) -> bool + Send + Sync + 'static>,
+    make_writer: W,
+    fmt_fields: N,
+    fmt_event: E,
+    _inner: PhantomData<S>,
+}
+
+impl<S, N, E, W> fmt::Debug for FmtLayer<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FmtLayer").finish()
+    }
+}
+
+/// A builder for `FmtLayer` that logs formatted representations of `tracing` events.
+pub struct FmtLayerBuilder<
+    S = Registry,
+    N = format::DefaultFields,
+    E = format::Format<format::Full>,
+    W = fn() -> io::Stdout,
+> {
+    fmt_fields: N,
+    fmt_event: E,
+    make_writer: W,
+    is_interested: Box<dyn Fn(&Event<'_>) -> bool + Send + Sync + 'static>,
+    _inner: PhantomData<S>,
+}
+
+impl<S, N, E, W> fmt::Debug for FmtLayerBuilder<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FmtLayerBuilder").finish()
+    }
+}
+
+impl FmtLayer {
+    /// Creates a [FmtLayerBuilder].
+    pub fn builder() -> FmtLayerBuilder {
+        FmtLayerBuilder::default()
+    }
+}
+
+impl<S, N, E, W> FmtLayerBuilder<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    /// Sets a filter for events. This filter applies to this, and
+    /// subsequent, layers.
+    pub fn with_interest<F>(self, f: F) -> Self
+    where
+        F: Fn(&Event<'_>) -> bool + Send + Sync + 'static,
+    {
+        FmtLayerBuilder {
+            fmt_fields: self.fmt_fields,
+            fmt_event: self.fmt_event,
+            make_writer: self.make_writer,
+            is_interested: Box::new(f),
+            _inner: self._inner,
+        }
+    }
+}
+
+// This, like the MakeWriter block, needs to be a seperate impl block because we're
+// overriding the `E` type parameter with `E2`.
```

Comment:
I don't think this is true...I think these can all go in the same impl block?

Reply:
I don't think they _can_. Moving to a separate impl block was the only thing that allowed me to do this re-assignment. For what its worth, `FmtSubscriber` does the same thing.

Reply:
I don't believe that's true. The subscriber builder had separate impl blocks where different _bounds_ on the type parameter are necessary. A method that returns a new value with a different type parameter can go in the same impl block as other methods.

For example, this impl block contains _several_ methods that change a generic type parameter:
https://github.com/tokio-rs/tracing/blob/2174dc243246b1c87a55dfb4602769b6732aed11/tracing-subscriber/src/fmt/mod.rs#L429-L667